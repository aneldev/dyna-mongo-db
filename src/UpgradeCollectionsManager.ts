import {DynaMongoDB} from "./index";
import {
  Collection,
  Db,
} from "mongodb";
import {
  dynaError,
  IDynaError,
} from "dyna-error";
import {DynaJobQueue} from "dyna-job-queue";

export interface IUpgradeCollectionsManagerConfig {
  dmdb: DynaMongoDB;
  upgradeCollections: ICollectionsUpgrades;
  onUpgradeError?: (collectionName: string, version: number, error: any) => void;
}

export interface ICollectionsUpgrades {
  [collectionName: string]: ICollectionUpgrades;
}

export interface ICollectionUpgrades {
  upgrades: ICollectionUpgrade[];
}

export interface ICollectionUpgrade {
  version: number;
  title: string;
  description?: string;
  method: (params: { db: Db; collectionName: string }) => Promise<void>;
}

interface IDBCollectionVersionInfo {
  collectionName: string;
  version: number;
}

export interface IUpgradeCollectionResults {
  initialVersion: number | null;
  upgradeToVersion: number | null;
  hasUpgrades: boolean | null;
  plannedUpgrades: number;
  appliedUpgrades: number;
}

const COLLECTION_VERSIONS_COLLECTION_NAME = 'dyna-mongo-db--upgrade-manager';

export class UpgradeCollectionsManager {
  private readonly dmdb = this.config.dmdb;

  constructor(private readonly config: IUpgradeCollectionsManagerConfig) {
    this.checkUpgradeVersions(config.upgradeCollections);
  }

  public addCollectionsUpgrades(collectionsUpgrades: ICollectionsUpgrades): void {
    this.checkUpgradeVersions(collectionsUpgrades);
    this.config.upgradeCollections = {
      ...this.config.upgradeCollections,
      ...collectionsUpgrades,
    };
  }

  public async upgradeCollection(collectionName: string): Promise<IUpgradeCollectionResults> {
    const output: IUpgradeCollectionResults = {
      initialVersion: null,
      upgradeToVersion: null,
      hasUpgrades: null,
      plannedUpgrades: 0,
      appliedUpgrades: 0,
    };
    const asCollectionName: string =
      collectionName.includes('---')
        ? collectionName.split('---')[1]
        : collectionName;

    const collectionVersion = await this.getCollectionVersion(collectionName);
    output.initialVersion = collectionVersion === -2 ? null : collectionVersion;

    const upgradeCollection = this.config.upgradeCollections[asCollectionName];
    output.hasUpgrades = !!upgradeCollection;
    if (!upgradeCollection) return output;

    return new Promise<IUpgradeCollectionResults>((resolve: (output: IUpgradeCollectionResults) => void, reject: (error: IDynaError) => void): void => {
      const queue = new DynaJobQueue();
      let ok = true;
      let error: IDynaError;

      const collectionMissingVersions =
        upgradeCollection.upgrades
          .sort((a, b) => a.version - b.version)
          .filter(upgradeCollection => upgradeCollection.version > collectionVersion);

      collectionMissingVersions.forEach(upgrade => {
        output.upgradeToVersion = upgrade.version;
        output.plannedUpgrades++;
        queue.addJobPromised(async () => {
          if (!ok) return;
          try {
            await this.checkAndUpgradeCollection(collectionName, upgrade);
            output.appliedUpgrades++;
          }
          catch (e) {
            ok = false;
            error = e;
          }
        });
      });
      queue.addJobPromised(async () => {
        if (ok) resolve(output);
        reject(error);
      });
    });
  }

  private async checkAndUpgradeCollection(collectionName: string, upgrade: ICollectionUpgrade): Promise<void> {
    console.log(`DynaMongoDB: Upgrade collection "${collectionName}" to version ${upgrade.version} ${upgrade.title}`);
    try {
      const db = await this.dmdb.getDb();
      await upgrade.method({
        db,
        collectionName,
      });
      await this.bumpConnectionCurrentVersion(collectionName, upgrade.version);
      console.log(`DynaMongoDB:  SUCCESS upgrade for collection "${collectionName}" to version ${upgrade.version}`);
    }
    catch (error) {
      const errorMessage = `DynaMongoDB: FAILED upgrade for collection "${collectionName}" to version ${upgrade.version}`;
      console.error(errorMessage, error);
      if (this.config.onUpgradeError) this.config.onUpgradeError(collectionName, upgrade.version, error);
      throw dynaError({
        code: 202206230850,
        message: errorMessage,
        parentError: error,
      });
    }
  }

  public async getCollectionVersion(collectionName: string): Promise<number> {
    const db = await this.dmdb.getDb();

    const versionCollection =
      await db
        .collection(COLLECTION_VERSIONS_COLLECTION_NAME)
        .findOne<IDBCollectionVersionInfo>({collectionName});

    if (versionCollection) return versionCollection.version;
    return -2;
  }

  private async getVersionsCollectionCollection(): Promise<Collection<IDBCollectionVersionInfo>> {
    const db = await this.dmdb.getDb();
    let versionsCollectionsCollection: Collection<IDBCollectionVersionInfo>;
    const collectionExists = await this.dmdb.collectionExists(COLLECTION_VERSIONS_COLLECTION_NAME);

    if (collectionExists) {
      // Load
      versionsCollectionsCollection = await db.collection<IDBCollectionVersionInfo>(COLLECTION_VERSIONS_COLLECTION_NAME);
    }
    else {
      // Create
      versionsCollectionsCollection = await db.createCollection<IDBCollectionVersionInfo>(COLLECTION_VERSIONS_COLLECTION_NAME);
      await versionsCollectionsCollection.createIndex(
        {collectionName: 1},
        {
          name: 'Collection name index',
          unique: true,
        },
      );
    }
    return versionsCollectionsCollection;
  }

  private async bumpConnectionCurrentVersion(collectionName: string, toVersion: number): Promise<void> {
    const versionsCollectionsCollection = await this.getVersionsCollectionCollection();
    const updateResult = await versionsCollectionsCollection
      .updateOne(
        {collectionName: collectionName} as IDBCollectionVersionInfo,
        {$set: {version: toVersion}},
        {upsert: true},
      );
    if (updateResult.upsertedCount === 0 && updateResult.modifiedCount === 0) {
      throw dynaError({
        message: `Cannot update version info collection for collection [${collectionName}] v${toVersion}`,
        data: {
          collectionName,
          updateResult,
        },
      });
    }
  }

  public async dropCollection(collectionName: string): Promise<void> {
    const versionsCollectionsCollection = await this.getVersionsCollectionCollection();
    await versionsCollectionsCollection
      .deleteOne(
        {collectionName: collectionName} as IDBCollectionVersionInfo,
      );
  }

  private checkUpgradeVersions(collectionsUpgrades: ICollectionsUpgrades): void {
    const errors: string[] = [];
    Object.keys(collectionsUpgrades)
      .forEach(collectionName => {
        const collectionUpgrades = collectionsUpgrades[collectionName];
        collectionUpgrades.upgrades.forEach((collectionUpgrade, index, arr) => {
          const prev = arr[index - 1];
          if (!prev) return;
          if (collectionUpgrade.version <= prev.version) {
            errors.push(`DynaMongoDB: error: 202206230915: Upgrade script for collection [${collectionName}] "${collectionUpgrade.title}" version (${collectionUpgrade.version}) has lower or same version with the previous upgrade script!!! Version numbers should be sequential!!!`);
          }
        });
      });
    if (errors.length) {
      errors.forEach(e => console.error(e));
      throw dynaError({
        code: 202206230915,
        message:
          errors.length === 1
            ? errors[0]
            : `DynaMongoDB: error: 202206230915: There are ${errors.length} Upgrade scripts with wrong versions!!!`,
        data: {errors},
      });
    }
  }

  public async _debug_changeVersion(collectionName: string, version: number): Promise<void> {
    const versionsCollection = await this.getVersionsCollectionCollection();
    await versionsCollection.updateOne(
      {collectionName},
      {$set: {version}},
    );
  }
}
