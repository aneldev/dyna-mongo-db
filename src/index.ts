export {
  DynaMongoDB,
  IDynaMongoDBConfig,
  IDatabaseUpgrade,
} from "./DynaMongoDB";

export {
  ICollectionsUpgrades,
  ICollectionUpgrades,
  ICollectionUpgrade,
} from "./UpgradeCollectionsManager";

export {
  saveDoc,
  loadDoc,
} from "./utils/saveLoadDoc";

export {
  removeId,
} from "./utils/removeId";

export {
  $setData,
} from "./utils/$setData";
