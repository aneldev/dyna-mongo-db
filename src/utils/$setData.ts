import {dynaObjectScan} from "dyna-object-scan";
import {SetFields} from "mongodb";

export const $setData = <TSchema = any, >(data: TSchema, propertyName?: string): SetFields<TSchema> | Partial<TSchema> => {
  let output = {};
  dynaObjectScan(data, (
    {
      path,
      value,
      parent,
      skip,
    },
  ) => {
    const isUndefined = value === undefined;
    if (isUndefined) return;

    const valueHasMongodbProperties =
      typeof value === "object" &&
      value !== null &&
      Object.keys(value).length &&
      Object.keys(value)
        .reduce((acc: boolean, key) => acc && key[0] === '$', true);
    const isRoot = parent === undefined;
    const isNull = value === null;
    const isArray = Array.isArray(value);
    const isObject = typeof value === "object" && !isNull && !isArray;

    const isObjectToOverwrite = isObject && value.__overwrite === true;
    const isArrayToOverwrite = isArray && value[value.length - 1] === '__overwrite';

    const applyPropertyName =
      (
        propertyName
          ? propertyName + path
          : (path || '').substr(1)
      )
        .replace(/\[/g, '.')    // Remove [
        .replace(/\]\./g, '.')  // Remove ].
        .replace(/\]/g, '.')    // Remove ]
        .replace(/\.$/, "");    // Remove ending .

    const setValue = (value: any): void => {
      if (isRoot) {
        if (propertyName) {
          output = {[propertyName]: value};
        }
        else {
          output = value;
        }
      }
      else {
        output[applyPropertyName] = value;
      }
    };

    if (valueHasMongodbProperties) {
      setValue(value);
      skip();
      return;
    }

    if (isNull) {
      setValue(value);
      return;
    }

    if (isObjectToOverwrite) {
      const applyValue = {...value};
      delete applyValue.__overwrite;
      setValue(applyValue);
      skip();
      return;
    }
    if (isObject && !isArray) {
      return;
    }

    if (isArrayToOverwrite) {
      setValue(value.concat().slice(0, -1));
      skip();
      return;
    }
    if (isArray) {
      return;
    }

    setValue(value);
  });

  return output;
};
