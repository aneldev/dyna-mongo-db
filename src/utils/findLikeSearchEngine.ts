export const findLikeSearchEngine = (fieldName: string, searchText?: string) => {
  if (!searchText || !searchText.trim()) return {};
  const searchParts = searchText.split(' ').filter(Boolean);
  return {
    $and: searchParts
      .map(searchPart => {
        const isNot = searchPart[0] === '-';
        const text = isNot ? searchPart.substr(1) : searchPart;
        const comparison = {
          $regex: `.*${text}.*`,
          $options: 'i',
        };
        return isNot ?
          {[fieldName]: {$not: comparison}}
          : {[fieldName]: comparison};
      }),
  };
};
