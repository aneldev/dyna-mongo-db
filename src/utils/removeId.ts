export const removeId = <TData, >(data: TData): Omit<TData, "id"> => {
  const data_: any = {...data};
  delete data_._id;
  return data_;
};
