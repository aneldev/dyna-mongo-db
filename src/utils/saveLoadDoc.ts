import {ObjectId} from "mongodb";

export const saveDoc = <TData, >(item: TData): any => {
  const item_: any = {...item};
  if (item_.id !== undefined) {
    if (item_.id) item_._id = new ObjectId(item_.id);
    delete item_.id;
  }
  return item_;
};

export const loadDoc = <TData, >(item: TData): TData => {
  const item_: any = {...item};
  item_.id = item_._id.toHexString();
  delete item_._id;
  return item_;
};
