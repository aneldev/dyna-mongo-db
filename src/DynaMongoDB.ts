import {Db, MongoClient, ObjectId, Collection, FilterQuery, SortOptionObject, CursorResult} from "mongodb";
import {UpgradeCollectionsManager, ICollectionsUpgrades, IUpgradeCollectionResults} from "./UpgradeCollectionsManager";

export interface IDynaMongoDBConfig {
  connectionString: string;
  databaseName: string;
  upgradeDatabase?: IDatabaseUpgrade[];
  upgradeCollections?: ICollectionsUpgrades;
  onUpgradeError?: (collectionName: string, version: number, error: any) => void;
}

export interface IDatabaseUpgrade {
  version: number;
  title: string;
  description?: string;
  method: (params: { db: Db }) => Promise<void>;
}

export interface IDynaMongoDBExplain {
  mongoDBExplain: CursorResult;
  usedIndex?: string;
  usedIndexName?: string;
}

export class DynaMongoDB {
  private db: Db | null = null;
  private mongoClient: MongoClient | null = null;
  private readonly upgradeCollectionsManager: UpgradeCollectionsManager;
  private collectionsCache: { [collectionName: string]: Collection<any> } = {};

  constructor(private readonly config: IDynaMongoDBConfig) {
    this.upgradeCollectionsManager = new UpgradeCollectionsManager({
      dmdb: this,
      upgradeCollections: {
        '@@dyna-mongo-db--database': {
          upgrades: this.config.upgradeDatabase || [],
        },
        ...(this.config.upgradeCollections || {}),
      },
      onUpgradeError: this.config.onUpgradeError,
    });
  }

  // DB Connection / Disconnection

  public async connect(): Promise<Db> {
    const {
      connectionString,
      databaseName,
    } = this.config;

    this.mongoClient = await MongoClient.connect(
      encodeURI(connectionString),
      {
        useNewUrlParser: true,
        useUnifiedTopology: true,
      },
    );

    this.db = this.mongoClient.db(databaseName);

    await this.upgradeDatabase();

    return this.db;
  }

  public async getDb(): Promise<Db> {
    if (this.db) return this.db;
    return this.connect();
  }

  public async disconnect(): Promise<void> {
    if (!this.mongoClient) return;
    if (!this.db) return;
    this.collectionsCache = {};
    await this.mongoClient.close();
    this.mongoClient = null;
    this.db = null;
  }

  public async reconnect(): Promise<Db> {
    await this.disconnect();
    return this.connect();
  }

  // General tools

  public ObjectId = ObjectId;

  // Collection tools

  public async createCollection<TSchema>(collectionName: string): Promise<Collection<TSchema>> {
    const db = await this.getDb();
    const collection = await db.createCollection<TSchema>(collectionName);
    await this.upgradeCollectionsManager.upgradeCollection(collectionName);
    this.collectionsCache[collectionName] = collection;
    return collection;
  }

  public async getCollectionNames(): Promise<string[]> {
    const db = await this.getDb();
    return new Promise((resolve, reject) => {
      db.listCollections().toArray((err, collections) => {
        if (err)
          reject(err);
        else
          resolve(
            collections
              .map(collectionInfo => collectionInfo.name),
          );
      });
    });
  }

  public async collectionExists(collectionName: string): Promise<boolean> {
    const collectionNames = await this.getCollectionNames();
    return collectionNames.indexOf(collectionName) > -1;
  }

  public async getCollection<TSchema>(collectionName: string): Promise<Collection<TSchema>> {
    const db = await this.getDb();
    const cachedCollection = this.collectionsCache[collectionName];
    if (cachedCollection) return cachedCollection;
    await this.upgradeCollectionsManager.upgradeCollection(collectionName);
    const collection = db.collection<TSchema>(collectionName);
    this.collectionsCache[collectionName] = collection;
    return collection;
  }

  public async dropCollection(collectionName: string): Promise<boolean> {
    const db = await this.getDb();
    await this.upgradeCollectionsManager.dropCollection(collectionName);
    try {
      await db.dropCollection(collectionName);
      delete this.collectionsCache[collectionName];
      return true;
    } catch (e) {
      if (e.name = 'MongoError' && e.code === 26 && e.message === 'ns not found') return false;
      throw e;
    }
  }

  public getCollectionVersion(collectionName: string): Promise<number> {
    return this.upgradeCollectionsManager.getCollectionVersion(collectionName);
  }

  // Upgrade methods

  public addCollectionsUpgrades(collectionsUpgrades: ICollectionsUpgrades): void {
    this.upgradeCollectionsManager.addCollectionsUpgrades(collectionsUpgrades);
  }

  public upgradeDatabase(): Promise<IUpgradeCollectionResults> {
    return this.upgradeCollectionsManager.upgradeCollection('@@dyna-mongo-db--database');
  }

  public upgradeCollection(collectionName: string): Promise<IUpgradeCollectionResults> {
    return this.upgradeCollectionsManager.upgradeCollection(collectionName);
  }

  // Document tools

  public async findFirst<TSchema = any>(
    {
      collectionName,
      filter = {},
      sort = {},
    }
      : {
      collectionName: string;
      filter?: FilterQuery<TSchema>;
      sort?: SortOptionObject<TSchema>; // MongoDB does not guarantee the order of query results, you should order them
    },
  ): Promise<TSchema | null> {
    const items = await this.find<TSchema>({
      collectionName,
      filter,
      sort,
      limit: 1,
    });
    return items[0] || null;
  }

  public async find<TSchema = any>(
    {
      collectionName,
      filter = {},
      sort = {},
      limit,
    }
      : {
      collectionName: string;
      filter?: FilterQuery<TSchema>;
      sort?: SortOptionObject<TSchema>; // MongoDB does not guarantee the order of query results, you should order them
      limit?: number;
    },
  ): Promise<TSchema[]> {
    const collection = await this.getCollection<TSchema>(collectionName);
    return limit === undefined
      ? collection.find<TSchema>(filter).sort(sort).toArray()
      : collection.find<TSchema>(filter).sort(sort).limit(limit).toArray();
  }

  public async explain<TSchema = any>(
    {
      collectionName,
      filter = {},
      sort = {},
      limit,
    }
      : {
      collectionName: string;
      filter?: FilterQuery<TSchema>;
      sort?: SortOptionObject<TSchema>; // MongoDB does not guarantee the order of query results, you should order them
      limit?: number;
    },
  ): Promise<IDynaMongoDBExplain> {
    const collection = await this.getCollection<TSchema>(collectionName);
    const mongoExplain: any = await (
      limit === undefined
        ? collection.find<TSchema>(filter).sort(sort).explain()
        : collection.find<TSchema>(filter).sort(sort).limit(limit).explain()
    );

    return {
      mongoDBExplain: mongoExplain,
      usedIndex: mongoExplain?.queryPlanner?.winningPlan?.inputStage,
      usedIndexName: mongoExplain?.queryPlanner?.winningPlan?.inputStage?.indexName,
    };
  }

  public _debug_changeVersion(collectionName: string, version: number): Promise<void> {
    return this.upgradeCollectionsManager._debug_changeVersion(collectionName, version);
  }
}
