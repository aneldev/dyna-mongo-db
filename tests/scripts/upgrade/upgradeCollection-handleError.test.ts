import {dynaError} from "dyna-error";
import {DynaMongoDB} from "../../../src";
import {ICollectionsUpgrades} from "../../../src/UpgradeCollectionsManager";
import {testConnectionInfo} from "../../setup/testConnectionInfo";
import {hideConsoleError} from "../../utils/hideConsoleError";

const TEST_COLLECTION_NAME = 'test-48823-create-a-collection';
let willFail = true;

const collectionUpgrades: ICollectionsUpgrades = {
  [TEST_COLLECTION_NAME]: {
    upgrades: [
      {
        version: 10,
        title: 'Creation of the collection',
        method: async ({
          collectionName, db,
        }) => {
          await db.createCollection(collectionName);
        },
      },
      {
        version: 20,
        title: 'Add the first doc',
        method: async ({
          db, collectionName,
        }) => {
          if (willFail) throw dynaError({message: 'Test error'});
          const collection = await db.collection<any>(collectionName);
          await collection.insertOne({
            code: 1,
            info: 'My 1st doc',
          });
        },
      },
      {
        version: 30,
        title: 'Add the second doc',
        method: async ({
          db, collectionName,
        }) => {
          const collection = await db.collection<any>(collectionName);
          await collection.insertOne({
            code: 2,
            info: 'My 2nd doc',
          });
        },
      },
    ],
  },
};

describe('Upgrade Collections', () => {
  let dmdb: DynaMongoDB;
  const errors: any[] = [];

  const clearDb = async (): Promise<void> => {
    await dmdb.dropCollection(TEST_COLLECTION_NAME);
  };

  beforeAll(async () => {
    hideConsoleError(true);
    dmdb = new DynaMongoDB({
      connectionString: testConnectionInfo.connectionString,
      databaseName: testConnectionInfo.databaseName,
      upgradeCollections: collectionUpgrades,
      onUpgradeError: (collectionName, version, error) => errors.push({
        collectionName,
        version,
        error,
      }),
    });
    await dmdb.connect();
    await clearDb();
  });

  afterAll(async () => {
    hideConsoleError(false);
    await clearDb();
    await dmdb.disconnect();
  });

  describe('Create Collection', () => {
    it('Test collection should not exist', async () => {
      const exists = await dmdb.collectionExists(TEST_COLLECTION_NAME);
      expect(exists).toBe(false);
    });

    it('should fail trying to findFirst', async () => {
      expect.assertions(3);
      try {
        await dmdb.findFirst<any>({
          collectionName: TEST_COLLECTION_NAME,
          filter: {code: 2},
        });
      }
      catch (e) {
        expect(e.message).toBe('Test error');
        expect(errors).toMatchSnapshot();
        const collectionVersion = await dmdb.getCollectionVersion(TEST_COLLECTION_NAME);
        expect(collectionVersion).toBe(10);
      }
    });

    it('should fail again trying to findFirst', async () => {
      expect.assertions(3);
      try {
        await dmdb.findFirst<any>({
          collectionName: TEST_COLLECTION_NAME,
          filter: {code: 2},
        });
      }
      catch (e) {
        expect(e.message).toBe('Test error');
        expect(errors).toMatchSnapshot();
        const collectionVersion = await dmdb.getCollectionVersion(TEST_COLLECTION_NAME);
        expect(collectionVersion).toBe(10);
      }
    });

    it('Fix the error and try findFirst again', async () => {
      willFail = false;
      const doc = await dmdb.findFirst<any>({
        collectionName: TEST_COLLECTION_NAME,
        filter: {code: 2},
      });
      expect(doc).not.toBe(null);
      expect(doc.code).toBe(2);
      const collectionVersion = await dmdb.getCollectionVersion(TEST_COLLECTION_NAME);
      expect(collectionVersion).toBe(30);
    });

  });

});
