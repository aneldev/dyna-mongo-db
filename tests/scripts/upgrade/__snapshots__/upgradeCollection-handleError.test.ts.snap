// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Upgrade Collections Create Collection It should fail again trying to findFirst 1`] = `
Array [
  Object {
    "collectionName": "test-48823-create-a-collection",
    "error": [Error: Test error],
    "version": 20,
  },
  Object {
    "collectionName": "test-48823-create-a-collection",
    "error": [Error: Test error],
    "version": 20,
  },
]
`;

exports[`Upgrade Collections Create Collection It should fail trying to findFirst 1`] = `
Array [
  Object {
    "collectionName": "test-48823-create-a-collection",
    "error": [Error: Test error],
    "version": 20,
  },
]
`;
